frontend/packages/portal-frontend/src/correlationAnalysis/utilities/helper.ts

```diff
@@ -0,0 +1,79 @@
+export function transformAndGroupByDataset(
+  associationsData: {
+    correlation: number;
+    log10qvalue: number;
+    other_dataset_id: string;
+    other_dimension_given_id: string;
+    other_dimension_label: string;
+  }[],
+  compoundDoseLabel: string,
+  datasetLookup: Record<string, string>,
+  compoundDoseToDose: Map<string, string>
+) {
+  const grouped: Map<string, any[]> = new Map();
+
+  // Step 1: Group items by other_dataset_id
+  associationsData.forEach((item) => {
+    const group = grouped.get(item.other_dataset_id) || [];
+    group.push(item);
+    grouped.set(item.other_dataset_id, group);
+  });
+
+  // Step 2: For each group, sort by correlation and assign rank
+  const result: Record<string, any[]> = {};
+
+  // eslint-disable-next-line no-restricted-syntax
+  for (const [datasetId, items] of grouped.entries()) {
```

Comment:
I think I'm going to remove this ESLint rule so you won't have to disable it. It's discouraging you from using a `for...of` loop because it thinks that's a performance problem. But it's really not.